cpp/ql/src/Security/CWE/CWE-190/TaintedAllocationSize.ql

```diff
@@ -15,31 +15,73 @@ import cpp
 import semmle.code.cpp.security.TaintTracking
 import TaintedWithPath
 
-predicate taintedChild(Expr e, Expr tainted) {
-  (
-    isAllocationExpr(e)
-    or
-    any(MulExpr me | me.getAChild() instanceof SizeofOperator) = e
-  ) and
-  tainted = e.getAChild() and
-  tainted.getUnspecifiedType() instanceof IntegralType
+/**
+ * Holds if `e` is a simple arithmetic expression that might be used to
+ * compute the size of an allocation, and `f` is a child of `e`.
+ */
+predicate arithmeticChild(Expr e, Expr f) {
+  e.(AddExpr).getAChild() = f or
+  e.(MulExpr).getAChild() = f
+}
+
+/**
+ * Holds if `e` is an allocation with size expression `size`, or `e` is a call
+ * to a 'wrapper' function with that size, that returns an allocation.  `mid`
+ * is the allocation that's directly wrapped and `alloc` is the true allocation
+ * at the end of the chain.
+ */
+predicate allocExprOrIndirect(Expr e, Expr size, Expr mid, AllocationExpr alloc) {
+  isAllocationExpr(alloc) and
+  size = alloc.getSizeExpr() and
```

Comment:
Can we reduce the false postive rate even more by excluding the allocations with a size we can prove will never overflow (e.g., with `SimpleRangeAnalysis.qll`?)

I tried doing a (very naive) comparison test between this new version of the query with and without the use of range analysis, and it does exclude some results: https://lgtm.com/query/7744692820054815595/

I haven't verified whether those results are meaningful yet, though.

Reply:
Thanks for this suggestion, I'll come back to it after the milestone.